### PR TITLE
[EAG2-28] Create Game

### PR DESCRIPTION
**Pull request by RayWayy**
Tests for create game and implementation.

### CODE REVIEW

File: src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java [hunk 1]
```diff
@@ -17,12 +18,13 @@
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Function;
 
-public class PersistentGame implements Game, SynchronousGame {
+public class 	PersistentGame implements Game, SynchronousGame {
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java]: please remove extra blank spaces

File: src/main/java/com/eleks/academy/whoami/model/response/PlayerWithState.java [hunk 2]
```diff
@@ -19,4 +19,9 @@ public class PlayerWithState {
 
 	private PlayerState state;
 
+	public PlayerWithState(SynchronousPlayer player) {
+		this.player = player;
+		this.state = PlayerState.READY;
```
Comment (VolodymyrKr-Eleks) [inline src/main/java/com/eleks/academy/whoami/model/response/PlayerWithState.java]: suggest you not hardcode, but add state as a parameter to the constructor and in a place where you create an instance put `PlayerState.READY` as a value

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	private final NewGameRequest gameRequest = new NewGameRequest();
+
+	@BeforeEach
+	public void setMockMvc() {
+		gameRequest.setMaxPlayers(5);
+	}
+
+	@Test
+	void save() {
+		final String player = "player";
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java:28]: suggest you look at the test one more time in the title of the test is written `GameInMemoryRepositoryTest`, but I can't find any usage or test of `GameInMemoryRepository.class` and his methods

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+	private final NewGameRequest gameRequest = new NewGameRequest();
+
+	@BeforeEach
+	public void setMockMvc() {
+		gameRequest.setMaxPlayers(4);
+	}
+
+	@Test
+	void createGame() {
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: please stick to the convention if it is a test that means that the test method should be named `createGameTest`

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+		final String player = "player";
+		SynchronousGame game = new PersistentGame(player, gameRequest.getMaxPlayers());
+
+		when(gameRepository.save(any(SynchronousGame.class))).thenReturn(game);
+
+		var createdGame = gameService.createGame(player, gameRequest);
+
+		assertNotNull(createdGame);
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: here you have checks(assertions) that value if not null, but much better to compare the result as an object with the object of the same type and values used for comparison. Library assertJ(should be already included) has the method `assertThat()` using it we can compare results with expectation in following way:
```
assertThat(gameService.createGame(player, gameRequest)).isEqualTo(new PersistentGame("player", 4))
```

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		assertNotNull(createdGame.getId());
+		assertNotNull(createdGame.getStatus());
+		assertNotNull(createdGame.getPlayers());
+		assertNull(createdGame.getCurrentTurn());
+
+		verify(gameRepository, times(1)).save(any(SynchronousGame.class));
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java:68]: is redundant if the mock method hasn't been used it won't return any data and the test will fail

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

File: src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java (hunk 1 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/core/impl/PersistentGame.java]: Resolved

File: src/main/java/com/eleks/academy/whoami/model/response/PlayerWithState.java (hunk 2 above)
Comment (RayWayy) [reply src/main/java/com/eleks/academy/whoami/model/response/PlayerWithState.java]: Resolved

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java (hunk 4 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: Resolved
Let me clarify. That mean I should add "test" word to all methods in tests? Like `GameControllerTest` and `GameInMemoryRepositoryTest`.

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java (hunk 6 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java:68]: Resolved
I'm not sure, but I think it is about only 51 line.

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java (hunk 5 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: Resolved
But with a few changes. I need the same type.

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java (hunk 3 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java:28]: Resolved
But I am not clearly sure if it is correct

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java [hunk 7]
```diff
@@ -3,47 +3,39 @@
...
-		when(games.put(any(), any())).thenReturn(game);
-
-		var savedGame = this.games.put(game.getId(), game);
-
-		assertNotNull(savedGame);
-		assertNotNull(savedGame.getId());
-		assertNotNull(savedGame.getStatus());
-		assertNotNull(savedGame.getPlayersInGame());
+		assertThat(gameRepository.save(game))
+				.isEqualTo(gameRepository.save(new PersistentGame("player", 4)));
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java]: just `.isEqualTo(new PersistentGame("player", 4))` you need to compare actual result with expected, but not 2 actual, if you do so you tests do not have sense

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java [hunk 8]
```diff
@@ -36,21 +34,22 @@ public void setMockMvc() {
...
 
 		when(gameRepository.save(any(SynchronousGame.class))).thenReturn(game);
 
 		var createdGame = gameService.createGame(player, gameRequest);
+		NewGameRequest gameRequest1 = new NewGameRequest();
+		gameRequest1.setMaxPlayers(4);
 
-		assertNotNull(createdGame);
+		assertThat(gameService.createGame(player, gameRequest))
+				.isEqualTo(gameService.createGame("player", gameRequest1));
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: here is the same `.ignoringFields("id").isEqualTo(GameDetails.builder()<fields to be set>.build()))`

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java (hunk 8 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: Resolved

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java [hunk 9]
```diff
@@ -29,13 +28,14 @@ public void setMockMvc() {
...
 		final String player = "player";
+
 		SynchronousGame game = new PersistentGame(player, gameRequest.getMaxPlayers());
 
 		assertThat(gameRepository.save(game))
-				.isEqualTo(gameRepository.save(new PersistentGame("player", 4)));
-
-		verify(gameRepository, times(2)).save(any(SynchronousGame.class));
+				.isEqualTo(new PersistentGame(player, 4));
+		assertThat(gameRepository.save(game).getPlayersInGame().get(0))
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java]: please compare whole object, not just part of it

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java (hunk 7 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java]: Resolved

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java [hunk 10]
```diff
@@ -35,21 +45,33 @@ public void setMockMvc() {
...
+				.isEqualTo(expectedGame);
+
+		assertNotNull(game.getId());
+		assertEquals(game.getId(), gameDetails.getId());
 
-		assertThat(gameService.createGame(player, gameRequest))
-				.isEqualTo(gameService.createGame("player", gameRequest1));
-		assertNotNull(createdGame.getId());
-		assertNotNull(createdGame.getStatus());
+		assertNotNull(game.getStatus());
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: better to check with specific value, not just it is null

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java (hunk 9 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java]: Resolved

File: src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java (hunk 10 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/service/GameServiceImplTest.java]: Resolved

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java [hunk 11]
```diff
@@ -35,7 +35,7 @@ void saveTest() {
 
 		assertThat(gameRepository.save(game))
 				.isEqualTo(new PersistentGame(player, 4));
-		assertThat(gameRepository.save(game).getPlayersInGame().get(0))
-				.isEqualTo(new PersistentGame(player, 4).getPlayersInGame().get(0));
+		assertThat(gameRepository.save(game).getPlayersInGame())
```
Comment (VolodymyrKr-Eleks) [inline src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java]: I meant whole `PersistentGame` object like: `assertThat(gameRepository.save(game)).isEqualTo(new PersistentGame(player, 4))`

**Review by VolodymyrKr-Eleks — CHANGES_REQUESTED** [review]

File: src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java (hunk 11 above)
Comment (RayWayy) [reply src/test/java/com/eleks/academy/whoami/repository/GameInMemoryRepositoryTest.java]: Resolved

**Review by VolodymyrKr-Eleks — APPROVED** [review]

**Pull request merged**